graph2tac/loader/py_data_server.py

```diff
@@ -171,23 +171,67 @@ def cluster_to_graph(self, cluster):
             ], dtype = self._def_name_dtype)
         )
 
-    
+class Splitter:
+    def datapoint(self, d : Definition):
+        raise Exception("Not implemented")
+    def cluster(self, d : list[Definition]):
+        raise Exception("Not implemented")
+
+class SplitByHash(Splitter):
+    def __init__(self, split = (8,1,1), random_seed = 0):
+        self.data_server = None # to be set by data_server
+        self.split = split
+        self.random_seed = random_seed
+    def datapoint(self, d : Definition):
+        # to make it identical to vasily's loader
+        ident_64 = np.array(d.node.identity, dtype = np.uint64).item()
+        return get_split_label(ident_64, self.split, self.random_seed)
+    def cluster(self, d : list[Definition]):
+        return 0
+
+class SplitByFilePrefix(Splitter):
+    def __init__(self, *prefixes_per_label):
+        self.data_server = None # to be set by data_server
+        self.prefixes_per_label = []
+        for label, prefixes in enumerate(prefixes_per_label):
+            if isinstance(prefixes, str):
+                prefixes = [prefixes]
+            else:
+                prefixes = list(prefixes)
+            self.prefixes_per_label.append((label+1, prefixes))
+    def datapoint(self, d : Definition):
+        fname = self.data_server.graphid_to_fname[d.node.graph]
+        for label, prefixes in reversed(self.prefixes_per_label):
+            if any(fname.startswith(prefix) for prefix in prefixes):
+                return label
+        return 0
+    def cluster(self, d : list[Definition]):
+        return self.datapoint(d[0])
+
 class DataServer(AbstractDataServer):
     def __init__(self,
                  data_dir: Path,
                  max_subgraph_size,
-                 split: tuple[int, int, int] = (8, 1, 1),
+                 split = (8, 1, 1),
```

Comment:
Isn't it time for us to get rid of the concept of holdout-set? I think we have all agreed in the past that we will almost certainly not use this.

Reply:
Up to discussion. I don't mind too much allowing more than two split labels.

Reply:
I would at least change the default to `(9, 1, 0)`.